zerver/migrations/0143_realm_bots_allowed_for_non_admins.py

```diff
@@ -0,0 +1,37 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.6 on 2018-01-28 13:10
+from __future__ import unicode_literals
+
+from django.db import migrations, models
+from django.db.backends.postgresql_psycopg2.schema import DatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+def set_initial_value_for_bots_allowed_for_non_admins(
+        apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
+    realm_model = apps.get_model("zerver", "Realm")
+    realms = realm_model.objects.all()
+    for realm in realms:
+        if realm.create_generic_bot_by_admins_only:
+            realm.bots_allowed_for_non_admins = 2
+        else:
+            realm.bots_allowed_for_non_admins = 1
+        realm.save(update_fields=["bots_allowed_for_non_admins"])
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('zerver', '0142_userprofile_translate_emoticons'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='realm',
+            name='bots_allowed_for_non_admins',
+            field=models.PositiveSmallIntegerField(default=1),
+        ),
+        migrations.RunPython(set_initial_value_for_bots_allowed_for_non_admins),
```

Comment:
You should include a `reverse_code` reverse migration, so we can undo this migration if needed.  Grep to see how we do this for other migrations.

I think what that should do with the new 3rd state of "no bots" is put at the more restrictive policy that exists in the old code.

Reply:
@timabbott I'm not too much experienced with `reverse_code`, so as you explained it helps in reversing the migration. So, I want to know how we reverse a migration(Do we reverse it after reverting the commit? because I found from the source code of `RunPython` that it simply executes the `reverse_code` function.)
If this is the case then here is what I think should work. 
```
def reverse_code(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
    realm_model = apps.get_model("zerver", "Realm")
    realms = realm_model.objects.all()
    for realm in realms:
        if realm.bot_creation_policy == 1:  # BOT_CREATION_EVERYONE
            realm.create_generic_bot_by_admins_only = False
        else:
            realm.create_generic_bot_by_admins_only = True
```
Is this right?
Can't find any other similar instances where we reuse migration it removes so grepping it didn't help much here.

Reply:
Yeah, so the use case for these is if you upgrade to Zulip 1.8, and something breaks, and you want to go back to 1.7 for a day while you investigate.  It's most likely to be run very soon after the upgrade (generally hours at most).  

That implementation looks correct to me, though I'd use the variable name `Realm` for `realm_model` for better readability (that's what we do in most other migrations), since what you're getting is what `Realm` was at that point in the series of database migrations.